src/fs/file_system.rs

```diff
@@ -0,0 +1,289 @@
+//! Module for [FileSystem].
+
+use super::*;
+use crate::fs::dir_entry_iter::UefiDirectoryEntryIterator;
+use crate::fs::file_info::BoxedUefiFileInfo;
+use crate::fs::path::{Path, PathError};
+use alloc_api::string::{FromUtf8Error, String, ToString};
+use alloc_api::vec::Vec;
+use core::fmt;
+use core::fmt::{Debug, Formatter};
+use uefi::proto::media::file::FileType;
+use uefi::table::boot::ScopedProtocol;
+use uefi::{CString16};
+
+// #[derive(Debug, Clone, Ord, PartialOrd, Eq, PartialEq, Hash)]
+#[derive(Debug, Clone)]
+pub enum FileSystemError {
+    /// Can't open the root directory of the underlying volume.
+    CantOpenVolume,
+    /// The path is invalid because of the underlying [PathError].
+    IllegalPath(PathError),
+    /// The file or directory was not found in the underlying volume.
+    FileNotFound(String),
+    /// The path is existend but does not correspond to a directory when a directory was expected.
+    NotADirectory(String),
+    /// The path is existend but does not correspond to a file when a file was expected.
+    NotAFile(String),
+    /// Can't delete the file.
+    CantDeleteFile(String),
+    /// Error writing bytes.
+    WriteFailure,
+    /// Error flushing file.
+    FlushFailure,
+    /// Error reading file.
+    ReadFailure,
+    /// Can't parse file content as UTF-8.
+    Utf8Error(FromUtf8Error),
+    /// Could not open the given path.
+    OpenError(String),
+}
+
+impl From<PathError> for FileSystemError {
+    fn from(err: PathError) -> Self {
+        Self::IllegalPath(err)
+    }
+}
+
+/// Return type for public [FileSystem] operations.
+pub type FileSystemResult<T> = Result<T, FileSystemError>;
+
+/// Entry point into the file system abstraction described by the module description. Accessor to
+/// an UEFI volume. Available methods try to be close to the `fs` module of `libstd`.
+///
+/// # Technical Background
+/// Some random interesting information how this abstraction helps.
+/// - To get metadata about UEFI files, we need to iterate the parent directory. This abstraction
+///   hides this complexity.
+pub struct FileSystem<'name, 'boot_services> {
+    /// Name to identify this file system. For example "root" or "boot". Only a help for users.
+    name: &'name str,
```

Comment:
I'm not sure how necessary the `name` field is, and under some circumstances it might be hard to provide a meaningful name (e.g. when iterating over all available file systems). If we dropped the name, these methods could move to be directly implemented on `SimpleFileSystem`.